hash.go

```diff
@@ -164,12 +170,31 @@ func hashWith(salt []byte, name string) string {
 	d.Write(salt)
 	d.Write(opts.Seed)
 	io.WriteString(d, name)
-	sum := nameBase64.EncodeToString(d.Sum(nil))
+	sum := make([]byte, nameBase64.EncodedLen(d.Size()))
+	nameBase64.Encode(sum, d.Sum(nil))
+	sum = sum[:length]
 
-	// TODO: Just make the first letter uppercase or lowercase as needed.
-	// This is also not needed for non-names, like import paths.
+	if !token.IsIdentifier(name) {
+		return string(sum)
+	}
+	// Keep the result a valid identifier which is equally exported or not.
+	if '0' <= sum[0] && sum[0] <= '9' {
+		// Turn "3foo" into "Dfoo".
+		sum[0] += 'A' - '0'
+	}
 	if token.IsExported(name) {
-		return "Z" + sum[:length]
+		if sum[0] == '_' {
+			// Turn "_foo" into "Zfoo".
+			sum[0] = 'Z'
+		} else if 'a' <= sum[0] && sum[0] <= 'z' {
+			// Turn "afoo" into "Afoo".
+			sum[0] -= 'a' - 'A'
+		}
+	} else {
+		if 'A' <= sum[0] && sum[0] <= 'Z' {
```

Comment:
just use `else if`

Reply:
Also thought about this initially, but it's on purpose. This way the upper/lower code is aligned with the other branch.